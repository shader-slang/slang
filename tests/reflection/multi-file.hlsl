//TEST:REFLECTION:-entry mainFS -profile ps_4_0 -target hlsl tests/reflection/multi-file-extra.hlsl -entry mainVS -profile vs_4_0

// Here we are testing the case where multiple translation units are provided
// at once, so that we want combined reflection information for the resulting
// program. The other part of this program is in `multi-file-extra.hlsl`.

float4 use(float  val) { return val; };
float4 use(float2 val) { return float4(val,0.0,0.0); };
float4 use(float3 val) { return float4(val,0.0); };
float4 use(float4 val) { return val; };
float4 use(Texture2D t, SamplerState s)
{
	// This is the vertex shader, so we can't do implicit-gradient sampling
	return t.SampleGrad(s, 0.0, 0.0, 0.0);
}

// Start with some parameters that will appear in both shaders
Texture2D sharedT;
SamplerState sharedS;
cbuffer sharedC
{
	float3 sharedCA;
	float  sharedCB;
	float3 sharedCC;
	float2 sharedCD;
}

// Then some parameters specific to this shader
// (these will get placed before the ones in the `extra` file,
// based on how they get named on the command-line)

Texture2D vertexT;
SamplerState vertexS;
cbuffer vertexC
{
	float3 vertexCA;
	float  vertexCB;
	float3 vertexCC;
	float2 vertexCD;
}

// And end with some shared parameters again
Texture2D sharedTV;
Texture2D sharedTF;


float4 mainFS() : SV_Target
{
	// Go ahead and use everything here, just to make sure things got placed correctly
	return use(sharedT, sharedS)
		+  use(sharedCD)
		+  use(vertexT, vertexS)
		+  use(vertexCD)
		+  use(sharedTV, vertexS)
		;
}