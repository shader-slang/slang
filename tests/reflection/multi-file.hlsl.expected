result code = 0
standard error = {
}
standard output = {
{
    "parameters": [
        {
            "name": "vertexT",
            "binding": {"kind": "shaderResource", "index": 0},
            "type": {
                "kind": "resource",
                "baseShape": "texture2D"
            }
        },
        {
            "name": "vertexS",
            "binding": {"kind": "samplerState", "index": 0},
            "type": {
                "kind": "samplerState"
            }
        },
        {
            "name": "vertexC",
            "binding": {"kind": "constantBuffer", "index": 0},
            "type": {
                "kind": "constantBuffer",
                "elementType": {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "vertexCA",
                            "type": {
                                "kind": "vector",
                                "elementCount": 3,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 0, "size": 12}
                        },
                        {
                            "name": "vertexCB",
                            "type": {
                                "kind": "scalar",
                                "scalarType": "float32"
                            },
                            "binding": {"kind": "uniform", "offset": 12, "size": 4}
                        },
                        {
                            "name": "vertexCC",
                            "type": {
                                "kind": "vector",
                                "elementCount": 3,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 16, "size": 12}
                        },
                        {
                            "name": "vertexCD",
                            "type": {
                                "kind": "vector",
                                "elementCount": 2,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 32, "size": 8}
                        }
                    ]
                },
                "containerVarLayout": {
                    "binding": {"kind": "constantBuffer", "index": 0}
                },
                "elementVarLayout": {
                    "type": {
                        "kind": "struct",
                        "fields": [
                            {
                                "name": "vertexCA",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 3,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 0, "size": 12}
                            },
                            {
                                "name": "vertexCB",
                                "type": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                },
                                "binding": {"kind": "uniform", "offset": 12, "size": 4}
                            },
                            {
                                "name": "vertexCC",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 3,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 16, "size": 12}
                            },
                            {
                                "name": "vertexCD",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 2,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 32, "size": 8}
                            }
                        ]
                    },
                    "binding": {"kind": "uniform", "offset": 0, "size": 40}
                }
            }
        },
        {
            "name": "fragmentT",
            "binding": {"kind": "shaderResource", "index": 1},
            "type": {
                "kind": "resource",
                "baseShape": "texture2D"
            }
        },
        {
            "name": "fragmentS",
            "binding": {"kind": "samplerState", "index": 1},
            "type": {
                "kind": "samplerState"
            }
        },
        {
            "name": "fragmentC",
            "binding": {"kind": "constantBuffer", "index": 1},
            "type": {
                "kind": "constantBuffer",
                "elementType": {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "fragmentCA",
                            "type": {
                                "kind": "vector",
                                "elementCount": 3,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 0, "size": 12}
                        },
                        {
                            "name": "fragmentCB",
                            "type": {
                                "kind": "scalar",
                                "scalarType": "float32"
                            },
                            "binding": {"kind": "uniform", "offset": 12, "size": 4}
                        },
                        {
                            "name": "fragmentCC",
                            "type": {
                                "kind": "vector",
                                "elementCount": 3,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 16, "size": 12}
                        },
                        {
                            "name": "fragmentCD",
                            "type": {
                                "kind": "vector",
                                "elementCount": 2,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 32, "size": 8}
                        }
                    ]
                },
                "containerVarLayout": {
                    "binding": {"kind": "constantBuffer", "index": 0}
                },
                "elementVarLayout": {
                    "type": {
                        "kind": "struct",
                        "fields": [
                            {
                                "name": "fragmentCA",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 3,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 0, "size": 12}
                            },
                            {
                                "name": "fragmentCB",
                                "type": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                },
                                "binding": {"kind": "uniform", "offset": 12, "size": 4}
                            },
                            {
                                "name": "fragmentCC",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 3,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 16, "size": 12}
                            },
                            {
                                "name": "fragmentCD",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 2,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 32, "size": 8}
                            }
                        ]
                    },
                    "binding": {"kind": "uniform", "offset": 0, "size": 40}
                }
            }
        },
        {
            "name": "sharedT",
            "binding": {"kind": "shaderResource", "index": 2},
            "type": {
                "kind": "resource",
                "baseShape": "texture2D"
            }
        },
        {
            "name": "sharedS",
            "binding": {"kind": "samplerState", "index": 2},
            "type": {
                "kind": "samplerState"
            }
        },
        {
            "name": "sharedC",
            "binding": {"kind": "constantBuffer", "index": 2},
            "type": {
                "kind": "constantBuffer",
                "elementType": {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "sharedCA",
                            "type": {
                                "kind": "vector",
                                "elementCount": 3,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 0, "size": 12}
                        },
                        {
                            "name": "sharedCB",
                            "type": {
                                "kind": "scalar",
                                "scalarType": "float32"
                            },
                            "binding": {"kind": "uniform", "offset": 12, "size": 4}
                        },
                        {
                            "name": "sharedCC",
                            "type": {
                                "kind": "vector",
                                "elementCount": 3,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 16, "size": 12}
                        },
                        {
                            "name": "sharedCD",
                            "type": {
                                "kind": "vector",
                                "elementCount": 2,
                                "elementType": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                }
                            },
                            "binding": {"kind": "uniform", "offset": 32, "size": 8}
                        }
                    ]
                },
                "containerVarLayout": {
                    "binding": {"kind": "constantBuffer", "index": 0}
                },
                "elementVarLayout": {
                    "type": {
                        "kind": "struct",
                        "fields": [
                            {
                                "name": "sharedCA",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 3,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 0, "size": 12}
                            },
                            {
                                "name": "sharedCB",
                                "type": {
                                    "kind": "scalar",
                                    "scalarType": "float32"
                                },
                                "binding": {"kind": "uniform", "offset": 12, "size": 4}
                            },
                            {
                                "name": "sharedCC",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 3,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 16, "size": 12}
                            },
                            {
                                "name": "sharedCD",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 2,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 32, "size": 8}
                            }
                        ]
                    },
                    "binding": {"kind": "uniform", "offset": 0, "size": 40}
                }
            }
        },
        {
            "name": "sharedTV",
            "binding": {"kind": "shaderResource", "index": 3},
            "type": {
                "kind": "resource",
                "baseShape": "texture2D"
            }
        },
        {
            "name": "sharedTF",
            "binding": {"kind": "shaderResource", "index": 4},
            "type": {
                "kind": "resource",
                "baseShape": "texture2D"
            }
        }
    ],
    "entryPoints": [
        {
            "name": "mainVS",
            "stage:": "vertex",
            "result:": {
                "semanticName": "SV_POSITION",
                "type": {
                    "kind": "vector",
                    "elementCount": 4,
                    "elementType": {
                        "kind": "scalar",
                        "scalarType": "float32"
                    }
                }
            }
        },
        {
            "name": "mainFS",
            "stage:": "fragment",
            "result:": {
                "stage": "fragment",
                "binding": {"kind": "varyingOutput", "index": 0},
                "semanticName": "SV_TARGET",
                "type": {
                    "kind": "vector",
                    "elementCount": 4,
                    "elementType": {
                        "kind": "scalar",
                        "scalarType": "float32"
                    }
                }
            }
        }
    ]
}
}
